Enable STREAMICE fixed-point treatment and other functionality with TAPENADE tool

dngoldberg opened the pull request:
## What changes does this PR introduce?
This introduces a number of modifications for the use of the Tapenade AD tool, primarily for use with the STREAMICE package but more generally to enable more flexible use of the tool. This addresses point 8 [issue 735](https://github.com/MITgcm/MITgcm/issues/735)


## What is the current behaviour? 
TAPENADE adjoint computation has been introduced to MITgcm, which works on several experiments including with the STREAMICE package. However, manual differentiation of the self adjoint linear system in STREAMICE has not been introduced nor has the low-memory fixed-point treatment for the nonlinear velocity solver that is implemented in OpenAD (https://gmd.copernicus.org/articles/9/1891/2016/). 

## What is the new behaviour 
1. Implementation of (fast) short-circuited adjoint of the STREAMICE matrix solve as in TAF and OpenAD. 
2. Implementation of the Christianson Fixed-point treatment for the STREAMICE nonlinear solve as in OpenAD.

## Does this PR introduce a breaking change? 
Due to the fixed point implementation, the tapenade halfpipe_streamice results file is modified.

## Other information:
above required significant refactoring of certain aspects of the STREAMICE package, introduction of a function (`streamice_tap_fixedpoint_notreduced`) to be called by Tapenade-generated code, and extensive changes to CPP directives withing STREAMICE files.

## Suggested addition to `tag-index`
o pkg/streamice, pkg/tapenade, tools/TAP_support
  - avoid differentiation of streamice_cg_solve with Tapenade, calling instead adstreamice_cg_solve (additional subroutines in pkg/tapenade/stubs_tap_adj, entry in tools/TAP_support/flow_tap
  - introduce fixed-point treatment of Goldberg et al 2016 (GMD) for streamice nonlinear solver using 
    Tapenade FP-LOOP pragma (breaks halfpipe_streamice tapenade experiment)
  - Additional source files to enable fixed-point differentation (tools/TAP_support/ADFirstAidKit/adFixedPoint.*, 
    pkg/streamice/streamice_tap_fixedpoint_notreduced.F

dngoldberg committed:
1st change to update tapenade capabilities for streamice (hand adjoint of cg solve, FP treatment)

dngoldberg committed:
more changes to implement FP loop

dngoldberg committed:
ensuring that verification runs and gives reasonable values

dngoldberg committed:
Merge branch 'master' into branch_tap_cg2d_fp

dngoldberg committed:
fixed point pragma on single line for residual decrease property

dngoldberg committed:
add logic to include streamice_cg_solve.f in tapenade command, only when streamice package is used

dngoldberg committed:
added flow commands in extra tap flow file

dngoldberg committed:
Merge branch 'master' into branch_tap_cg2d_fp

dngoldberg committed:
remove tap_dontcompile as in main branch

dngoldberg committed:
stashing changes

dngoldberg committed:
reverse popping of velocity arrays

dngoldberg committed:
new BWD s/rs for nocheckpoint mode, streamlining of calls to adstreamice_cg_solve to avoid recalculation, modifying fixed point convergence check, genmake2 updates to specify tapenade cmd line flags, linking libraries, and addl flags for adFixedpoint.c

dngoldberg committed:
before malking evern more chg:

dngoldberg committed:
Thwaites experiment runs with both tapenade and openad

dngoldberg committed:
more updates to ensure verification experiments work
(and are somewhat accurate on ARCHER)

dngoldberg committed:
modifying global sum call

dngoldberg committed:
remove matdestroy commands

dngoldberg committed:
make adjoint seed unique to proc 0

dngoldberg committed:
Merge branch 'master' into branch_tap_cg2d_fp

dngoldberg committed:
Revert output_tap_adj.txt.gz to previous state

dngoldberg committed:
Update genmake2

jm-c commented:
@dngoldberg I was about to take a look at this PR but you changed some files to "dos" format so I stopped there.
Could you revert these 3 files (adStack.c, adStack.h & exch_tap_b.F) to "unix" format ?

dngoldberg committed:
apply dos2unix on 3 files

dngoldberg commented:
> @dngoldberg I was about to take a look at this PR but you changed some files to "dos" format so I stopped there. Could you revert these 3 files (adStack.c, adStack.h & exch_tap_b.F) to "unix" format ?

Hi @jm-c That is strange, I did not work with these files on a windows platform. I have called "dos2unix" on all 3 files -- I hope this clears up the issue? let me know if you see dos characters anywhere else.

jm-c commented:
I added the "adjoint" label since I saw few OpenAD specific changes.
@dngoldberg Thanks for converting these 3 files, now the "diff" shows much less changes.
I also tried to run `testreport -tap -adm` on my laptop, but all experiments except `halfpipe_streamice` failed to compile:
```
usr/bin/ld: stubs_tap_adj.o: in function `streamice_cg_solve_fwd_':
stubs_tap_adj.f:(.text+0x4cc): undefined reference to `streamice_cg_solve_'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:1616: mitgcmuv_tap_adj] Error 1
```
And for `halfpipe_streamice`, it is running in about 3 min when using master branch but with this PR, it has been running for 144 min and still not done. How long does it takes when you run this test on your side ? (vs time to run master branch version) ?

dngoldberg committed:
fix issue with non-streamice tapenade; (temporarily) update results file for halfpipe_streamice

dngoldberg commented:
Hi @jm-c -- so sorry, i had forgotten to run other tests than `streamice`, i am so used to CI now. I have found and fixed the issue you were seeing with compilation -- at least, i have fixed it for global-ocean.cs32x15. it would take quite a while to run all tapenade tests, though I am certain all errors had the same cause which should now be fixed.

something is wrong if, with this branch, `halfpipe_streamice` takes 144+ minutes. In the `master` branch, the run is 328 seconds for the server/build I am using; it takes 260 **seconds** with the feature branch ie less than 5 min. (I have pushed the results file from my server to the branch, so you can see it, though i know it will be overwritten.)

I wonder if you are using an out of date installation of tapenade? there is no check in the testreport of this. Depending on when you downloaded it, Laurent Hascoet worked with my and I recall he made some changes to the binaries when i was testing the fixed-point treatment that is a part of the present PR. An option could be to use the docker image of [this PR](https://github.com/MITgcm/MITgcm/pull/931), thought I have not used this image for the present branch. Please let me know if you are unsure of your tapenade version and/or would like find the latest.

Dan

dngoldberg commented:
Apologies it looks like I forgot to push the commit. Will do in a few hours.

jm-c commented:
@dngoldberg Regarding `halfpipe_streamice`, I was using an old tapenade version (from: 2023-05-12); I switched to the latest (2025-07-29) and the problem is gone (run in 2.5 min).

jm-c committed:
Merge branch 'master' into branch_tap_cg2d_fp

jm-c commented:
@dngoldberg I think these comments fit better in this PR (rather than in PT #935):
> hi @jm-c im not sure how this will interact with https://github.com/MITgcm/MITgcm/pull/927, as i have manipulated the files passed to the tapenade tool. I did this as it was the suggested way of implementing a "hand written" adjoint: to pass the original file to the tapenade tool, but do not compile the resulting _b.f file. As a result, I have "manually" passed streamice_cg_solve.f to tapenade (when STREAMICE is enabled) rather than adding it to pkg/streamice/streamice_ad_diff.list which would not have the desired result. It might be OK with this PR, but it would be difficult to test.
> 
> The alternative method is to use the tools/TAP_support/flow_tap file (similar to TAF flow files). I have just tried this approach; i cannot recall why i did not use it before, but it does seem to work. I can make this change to the branch or leave it until you are able to give a detailed review.

so I am copying this here.

jm-c commented:
@dngoldberg I started to look at changes in `genmake2` and have few comments:

1. seems that the 3 new `genmake2` options ("-taparg", "-tapgcc", "-tapthead") are not necessary for the main purpose of this PR, i.e., to get the fixed-point method to work with `pkg/streamice`. I think it would be better to move this part in an other PR that would focus just on that (with example of what type of option I could specify as "-taparg", "-tapgcc" argument). Plus there is already some untested and old "pthread" implementation in MITgcm so might need to be careful not to add confusion.
2. regarding `streamice_cg_solve.f` special treatment, from what your wrote on Sep 21 (copied above), it seems a little bit like a hack. So if you have a working solution that use `flow_tap`, I think it would be better.
3. the addition of `adFixedPoint.c` looks good, no problem with that.

Also, regarding exp. `halfpipe_streamice` with Tapenade, the gradient-check are good for the adjoint run but not great for the tang.-lin test (RMS: 4.7). This was not the case with master branch (good grdchk, RMS: 7.E-6). Do you have an explanation for that ? Is it easy to fix ?

jm-c committed:
Merge branch 'master' into branch_tap_cg2d_fp

jm-c commented:
working on fixing the conflicts

jm-c committed:
Merge branch 'master' into branch_tap_cg2d_fp

jm-c committed:
minor cleaning

remove trailing blanks and use tab consistently

dngoldberg commented:
Hi J-M -- ok from looking at your commits i have the answer to my question. The changes you have made make sense.  My aim is to get rid of $XTRATAPFILE if I can. but I do not have a chance to do testing now -- i will do so when I do.

dngoldberg committed:
attempt: use tapenade flow file and set XTRATAPFILE to '' (to be removed) in genmake2

dngoldberg committed:
remove change to cost_dependent_init.F from this pull request

dngoldberg committed:
remove changes to genmake2, active_read/write_tap, exch_tap_b from this pull request, as well as non-streamice changes to stubs_tap_adj

dngoldberg committed:
add comment on commented-out function from L Hascoet

jm-c committed:
Merge branch 'master' into branch_tap_cg2d_fp

jm-c committed:
remove tabs and trailing blanks

jm-c reviewed (commented):
I don't know much about `pkg/streamice` and even less about Fix-Point method so
my review is going to be very superficial.
 
I think it's a good practice to clarify what CPP options are doing and how/where
to set them, since user may need to switch on/off theses. So I made some comments related to CPP options below.

And just a minor thing, which is not telated to this PR: may be this file and subroutine: `streamice_vel_solve_openad.F` could be renamed since it not specific to OpenAD ?

dngoldberg commented on pkg/streamice/STREAMICE_FP.h:
```diff
@@ -0,0 +1,51 @@
+     & STREAMICE_need2destroymat,
+     & STREAMICE_OAD_petsc_reuse,
+     & PETSC_PRECOND_TMP,
+     & PETSC_PRECOND_OAD
```
ok, i would prefer the second, when i was writing this it seemed a lot to have two separate commob blocks.

dngoldberg committed:
rename streamice_vel_solve_openad to streamice_vel_solve_fp

dngoldberg committed:
merging OAD_FP and TAP_FP cpp variables into a single ALLOW_STREAMICE_FP_ADJ

dngoldberg committed:
replacing fixed-point CPP variable in verification code folders

dngoldberg committed:
fixing issues from CPP replacement

dngoldberg committed:
replace cpp commamd in oad template

dngoldberg committed:
fixing CPP commands in add'l overlooked oad templates

dngoldberg committed:
split control blocks by variable type

dngoldberg committed:
Merge branch 'master' into branch_tap_cg2d_fp

jm-c commented on pkg/streamice/STREAMICE_FP.h:
```diff
@@ -0,0 +1,51 @@
+
+C STREAMICE_FP.h
+#ifdef ALLOW_OPENAD
+#include "STREAMICE_OPTIONS.h"
```
I tried to run the OpenAD `halfpipe_streamice` test, after commenting out this extra "#include "STREAMICE_OPTIONS.h"", both with the November 08 version and also with this PR current version, and the 2 tests compile and run fine. Could you provide an example where this extra include is needed for OpenAD ?
From my side, this looks suspicious since there are many other common block header files that are converted to module when using OpenAD without  the need for extra CPP header included.

dngoldberg committed:
Update STREAMICE_FP.h

Remove unneeded include

dngoldberg committed:
Update STREAMICE_OPTIONS.h

Comment explaining ALLOW_STREAMICE_FP_ADJ cpp variable

jm-c committed:
avoid unused variables

jm-c committed:
remove trailing blanks

and consecutive blank lines

jm-c committed:
reduce differences with version in pkg/streamice

jm-c committed:
update reference output:

- TAF-adm reference output from villon (Ubuntu)
- Tapenade-adj ref. output from baudelaire (FC.42)

jm-c commented:
@dngoldberg I updated 2 reference output (TAF and Tapenade Adj output) in `halfpipe_streanice`.
We need to decide what to do with `output_tap_tlm.txt` (i.e., the Tapenade Tang.Lin output). 3 options:
1. keep the current output (with good gradient-check) but this test will now fail badly.
2. update this reference output (with bad gradient-check) so that this test will pass on reference platform (but no guarantee elsewhere).
3. remove this reference output and drop this experiment from Tapenade Tang-Lin set of  testreport tests.

My view: I don't like (1);  we could go with (2) if we think there is a good chance this test will be fixed in the near future ; otherwise, because something is not right in the Tang-Lin grdchk, it might be very sensitive to anything and requiring an update of this output more often than normally expected. Option (3) would be my preference (we also don't have a tang-lin test with TAF for this experiment).

jm-c committed:
for now, disable Tapenade Tang-Lin test.

since gradient-check are not right.

jm-c committed:
remove obsolete (and commented out) lines

jm-c reviewed (approved):
LGTM

jm-c commented:
Will merge this PR later today. 
@dngoldberg And regarding the Tapenade TLM issue, from my side, there is no rush and since you (and the people you collaborate with) are the main users of `pkg/streamice` AD pieces, you are in a better position to decide what is the priority list.

jm-c committed:
Merge branch 'master' into branch_tap_cg2d_fp

jm-c committed:
document pkg/streamice Fixed-Point with Tapenade
